core/src/main/scala/org/apache/spark/SparkContext.scala

```diff
@@ -689,6 +689,61 @@ class SparkContext(config: SparkConf) extends Logging with ExecutorAllocationCli
     new ParallelCollectionRDD[T](this, seq, numSlices, Map[Int, Seq[String]]())
   }
 
+  /**
+   * Creates a new RDD[Long] containing elements from `start` to `end`(exclusive), increased by
+   * `step` every element.
+   *
+   * @note if we need to cache this RDD, we should make sure each partition does not exceed limit.
+   *
+   * @param start the start value.
+   * @param end the end value.
+   * @param step the incremental step
+   * @param numSlices the partition number of the new RDD.
+   * @return
+   */
+  def range(
+      start: Long,
+      end: Long,
+      step: Long = 1,
+      numSlices: Int = defaultParallelism): RDD[Long] = withScope {
+    assertNotStopped()
+    if (step == 0) {
+      // when step is 0, range will run infinite
+      throw new IllegalArgumentException("`step` cannot be 0")
+    }
+    val length: BigInt = {
+      val safeStart = BigInt(start)
+      val safeEnd = BigInt(end)
+      if (((safeEnd - safeStart) % step).toInt == 0) {
+        (safeEnd - safeStart) / step
+      } else {
+        (safeEnd - safeStart) / step + 1
+      }
+    }
+    parallelize(0 until numSlices, numSlices).mapPartitions(iter => {
+      val i = iter.next()
```

Comment:
we should use partition id here, not i. there is no guarantee that each partition has 1 number.